### PR TITLE
Batch append

### PR DESCRIPTION
**Pull request by George-Payne**
- Add batch append
 - decide if should batch append or append based off of support and credentials
 - update tests
  - http2-assertion-failure should use classic append
  - write-after-end should test both append styles (also fixes:  #215)
 - reconnect tests should use classic append

### CODE REVIEW

File: src/__test__/connection/reconnect.test.ts [hunk 1]
```diff
@@ -26,7 +26,9 @@ describe("reconnect", () => {
     // make successful append to connect to node
     const firstAppend = await client.appendToStream(
       "my_stream",
-      jsonEvent({ type: "first-append", data: { message: "test" } })
+      jsonEvent({ type: "first-append", data: { message: "test" } }),
+      // batch append triggers reconnect as soon as stream drops, so we need to force regular append
+      { credentials: { username: "admin", password: "changeit" } }
```
Comment (oskardudycz) [inline src/__test__/connection/reconnect.test.ts:31]: **QUESTION:** If someone uses explicit ACL params in append, then batch won't be used?

File: src/__test__/extra/write-after-end.test.ts [hunk 2]
```diff
@@ -1,39 +1,96 @@
...
+
+  optionalTest(matchServerVersion`>=21.10`)(
+    "Should not write after end (batch append)",
+    async () => {
+      const node = createTestNode();
+      await node.up();
+
+      const client = new EventStoreDBClient(
+        { endpoint: node.uri },
+        { rootCertificate: node.rootCertificate }
```
Comment (oskardudycz) [inline src/__test__/extra/write-after-end.test.ts:67]: **CONCERN:** Shouldn't we set global credentials here?

File: src/__test__/streams/appendToStream-batch-append.test.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  clientConstructer: any,
+  name: any,
+  _: any,
+  cache: any
+): Promise<Duplex> {
+  return this.GRPCStreamCreator(
+    clientConstructer,
+    name,
+    () => {
+      throw "Creater shouldn't be called as it will take the client from the cache";
```
Comment (oskardudycz) [inline src/__test__/streams/appendToStream-batch-append.test.ts:137]:
```suggestion
      throw "Creator shouldn't be called as it will take the client from the cache";
```

File: src/__test__/streams/appendToStream-batch-append.test.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+function spyOn(this: EventStoreDBClient, method: string) {
+  return jest.spyOn(this, method as never) as any;
+}
+
+function extractBatchStream(
+  this: EventStoreDBClient,
+  clientConstructer: any,
```
Comment (oskardudycz) [inline src/__test__/streams/appendToStream-batch-append.test.ts:128]:
```suggestion
  clientConstructor: any,
```

File: src/streams/appendToStream/index.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  event: EventData | EventData[],
+  {
+    expectedRevision = ANY,
+    batchAppendSize = 3 * 1024 * 1024,
+    ...baseOptions
+  }: AppendToStreamOptions = {}
+): Promise<AppendResult> {
+  const events = Array.isArray(event) ? event : [event];
+
+  if (
```
Comment (oskardudycz) [inline src/streams/appendToStream/index.ts:61]: See comment above: https://github.com/EventStore/EventStore-Client-NodeJS/pull/229/files#r741774656

File: src/__test__/extra/write-after-end.test.ts (hunk 2 above)
Comment (George-Payne) [reply src/__test__/extra/write-after-end.test.ts:67]: What's your reasoning?

File: src/streams/appendToStream/unpackError.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import {
+  AccessDeniedError,
+  MaxAppendSizeExceededError,
+  StreamDeletedError,
+  TimeoutError,
+  UnknownError,
+  WrongExpectedVersionError,
+} from "../..";
+
+export const unpackWrongExpectedVersion = (grpcError: Status) =>
```
Comment (oskardudycz) [inline src/streams/appendToStream/unpackError.ts:19]: **FIX:** I think we're missing tests for the logic in this file, so unpacking the wrong expected version. IMHO, we should add tests for negative scenarios and ensure that we have matching coverage to the regular append.

File: src/__test__/extra/write-after-end.test.ts (hunk 2 above)
Comment (oskardudycz) [reply src/__test__/extra/write-after-end.test.ts:67]: The other tests use credentials, so I think that we should be consistent. It'd be probably worth having a matrix for both calls with and without ACLs.

**Review by oskardudycz — COMMENTED** [review]
I'm not sure if selecting batch vs regular append based on credentials is not too magical (there is no RFC), but as it's the same as in .NET, then I could approve it when the test coverage is increased.
See comments above.

File: src/__test__/extra/write-after-end.test.ts (hunk 2 above)
Comment (George-Payne) [reply src/__test__/extra/write-after-end.test.ts:67]: > The other tests use credentials, so I think that we should be consistent.
Probably better not to be consistent in tests, as you're more likely to catch something when you do things differently?

> It'd be probably worth having a matrix for both calls with and without ACLs.

Credentials are specifically tested here:

https://github.com/EventStore/EventStore-Client-NodeJS/blob/master/src/__test__/connection/defaultCredentials.test.ts

This test file is ensuring that we don't have a regression on #208

Comment (oskardudycz) [reply src/__test__/extra/write-after-end.test.ts:67]: Inconsistency may be helpful, but better be consistent and proper matrix coverage 😉

Test that you linked checks only read methods it lacks checks for appends.

Comment (George-Payne) [reply src/__test__/extra/write-after-end.test.ts:67]: > Test that you linked checks only read methods it lacks checks for appends.

That's more a general hole in the testing, rather than this PR. I'll open an issue for it.

Comment (George-Payne) [reply src/__test__/extra/write-after-end.test.ts:67]: Issue is here #230

File: src/__test__/connection/reconnect.test.ts (hunk 1 above)
Comment (George-Payne) [reply src/__test__/connection/reconnect.test.ts:31]: As discussed:
- no explicit setting
- rewrite test not to use append, to avoid workaround

File: src/streams/appendToStream/index.ts (hunk 5 above)
Comment (George-Payne) [reply src/streams/appendToStream/index.ts:61]: For context:

In batch append, the stream was opened with default credentials, and we cannot add credentials to the specific append batch. Therefore, we fall back to a single append if per call credentials are passed.

File: src/__test__/connection/reconnect.test.ts (hunk 1 above)
Comment (oskardudycz) [reply src/__test__/connection/reconnect.test.ts:31]: 👍

File: src/streams/appendToStream/unpackError.ts (hunk 6 above)
Comment (George-Payne) [reply src/streams/appendToStream/unpackError.ts:19]: Additional tests for possible error types added here:
https://github.com/EventStore/EventStore-Client-NodeJS/pull/229/commits/22495039f75921504741c5a0825913e55860ece5

**Review by oskardudycz — APPROVED** [review]

**Review by thefringeninja — APPROVED** [review]
